File: src/org/fiz/RatingSection.java
/* Copyright (c) 2009 Stanford University
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package org.fiz;

/**
 * A RatingSection displays a series of rating images (stars, a progress bar, 
 * etc).  Users can select a rating by clicking on these images; upon clicking,
 * the selected rating is sent to a specified URL via Ajax.  When no rating has
 * been selected, a special "unrated" message is displayed.  RatingSections
 * support the following constructor properties:
 *   ajaxUrl:           (optional) A template for a URL to which an Ajax 
 *                      request will be sent every time a user selects a rating.
 *                      The template will be expanded using both the main
 *                      dataset and the result of any specified data request.  
 *                      The ClientRequest received by the {@code ajaxUrl} method
 *                      will have an additional "rating" key in its main 
 *                      dataset, with the associated value being the user's 
 *                      rating.  For instance, if {@code ajaxUrl} is set to 
 *                      "/x/ajaxY?a=b", then the dataset received by the method
 *                      for "ajaxY" will have {a:b, rating:[user rating]}.  Note
 *                      that the method (page) name must begin with "ajax".
 *   autoSizeText:      (optional) A boolean value determining whether or not 
 *                      the {@code unratedMessage}, the message shown when no 
 *                      rating has been selected, should be automatically sized 
 *                      to fit on one line passing through the center of the 
 *                      RatingSection.  If set as false, you may choose to 
 *                      customize the {@code <div>} containing the unrated 
 *                      message using CSS.  The unrated message text is located
 *                      in a {@code <div>} with class "unratedMsg", lying within
 *                      the RatingSection {@code <div>} whose id can be 
 *                      specified.  Thus, a CSS rule such as: 
 *                      <pre>
 *                          #[my-specified-id] .unratedMsg { ... }
 *                      </pre>
 *                      will give you CSS access to the unrated message text.
 *                      {@code autoSizeText} defaults to true.
 *   class:             (optional) Used as the {@code class} attribute for
 *                      the HTML div that contains the RatingSection.
 *   granularity:       (optional) Specifies the fineness of the rating 
 *                      measurements.  Ratings will be rounded to the nearest
 *                      granularity.  For instance, if {@code granularity} is 
 *                      set to 0.5, all ratings will be rounded to the nearest 
 *                      0.5.  If set to zero, no rounding will occur and the 
 *                      resulting ratings will be as fine-grained as the pixels.
 *                      {@code granularity} defaults to 1.0.  
 *   id:                (optional) Used as the id attribute for the {@code 
 *                      <div>} containing the RatingSection.
 *   imageFamily:       (optional) Specifies the partial URL of the images to 
 *                      display.  If {@code imageFamily} is specified as 
 *                      "/x/y/z.gif", then the following two images must exist:
 *                      - /x/y/z-on.gif:     The image to display when the
 *                                           rating value is selected.
 *                      - /x/y/z-off.gif:    The image to display when the 
 *                                           rating value is not selected.
 *                      Both of these images should have the same dimensions. 
 *                      Note that the size of the RatingSection will vary based
 *                      on the size of these images and the number of images 
 *                      specified - no image scaling will occur. {@code 
 *                      imageFamily} defaults to 
 *                      "/static/fiz/images/goldstar.png", images taken and/or
 *                      modified from the Crystal Icons Project.
 *   initRating:        (optional) Used as the initial rating for the 
 *                      RatingSection.  This value will be displayed when the 
 *                      RatingSection first loads.  If specified, any value 
 *                      for {@code initRatingKey} will be ignored.  If no 
 *                      initial rating is specified by either {@code initRating}
 *                      or {@code initRatingKey}, the RatingSection will 
 *                      initially display the {@code unratedMessage} and a
 *                      rating of zero.
 *   initRatingKey:     (optional) The dataset key to use for retrieval of an 
 *                      initial rating.  Both the ClientRequest's main dataset
 *                      and the DataRequest's response dataset will be searched.
 *                      For instance, if a {@code request} is specified with 
 *                      response containing {myrating : 3.2}, then setting 
 *                      {@code initRatingKey} to "myrating" will cause the 
 *                      initial rating to be 3.2.
 *   numImages:         (required) Specifies the number of rating images to 
 *                      display, and determines the maximum rating a user can 
 *                      select.  For instance, if the rating images are stars 
 *                      and {@code numImages} is set to 5, then five stars will
 *                      be displayed with the maximum user rating being 5.
 *   readOnly:          (optional) If set as true, then no user rating can be 
 *                      specified and the RatingSection will always show its 
 *                      initial rating.  {@code readOnly} defaults to false.
 *   request:           (optional) The name of a DataRequest whose result can 
 *                      supply data for any of the specified templates ({@code 
 *                      ajaxUrl}, {@code initRatingKey}, {@code 
 *                      unratedMessage}).  The request is created by the caller
 *                      and registered in the ClientRequest by calling 
 *                      ClientRequest.addDataRequest.
 *   unratedMessage:    (optional) A message to display on top of the 
 *                      RatingSection when no rating has been specified or 
 *                      selected.  {@code unratedMessage} defaults to "Not Yet 
 *                      Rated".
 *
 * RatingSection automatically sets the following {@code class} attributes
 * for use in CSS:
 *   offRating:         The {@code <div>} containing the "off" rating images.
 *   onRating:          The {@code <div>} containing the "on" rating images.
 *   unratedMsg:        The {@code <div>} containing the text of the {@code 
 *                      unratedMessage}.
 */
public class RatingSection extends Section implements DirectAjax {
    // The following are used as default values for various optional constructor
    // configurations.
    protected static final String DEFAULT_AJAX_URL = "";
    protected static final String DEFAULT_IMAGE_FAMILY = 
        "/static/fiz/images/goldstar24.png";
    protected static final String DEFAULT_INIT_RATING = "-1";
    protected static final String DEFAULT_UNRATED_MESSAGE = "Not Yet Rated";
    
    // Suffixes for files specified by the image family.
    protected static final String ON_SUFFIX = "-on";
    protected static final String OFF_SUFFIX = "-off";
    
    // A cache of the RatingSection's id.
    protected String id = null;
    
    /**
     * Construct a RatingSection.
     * @param properties            Contains configuration information for the 
     *                              RatingSection; see description above.
     */
    public RatingSection(Dataset properties) {
        this.properties = properties;
    }

    /**
     * Sets the rating of the RatingSection by emitting Javascript to that 
     * effect.  If the RatingSection has not already been rendered, no action 
     * will be taken.
     * 
     * @param cr                    Overall information about the client request
     *                              being serviced.
     * @param rating                The new rating for the RatingSection.
     */
    public void setRating(ClientRequest cr, double rating) {
        if (id == null) return;
        cr.evalJavascript("Fiz.ids.@1.setOfficialRating(@2);", id, "" + rating);
    }

    /**
     * Sets whether the RatingSection is read-only by emitting Javascript to 
     * that effect.  If the RatingSection has not already been rendered, no 
     * action will be taken.
     * 
     * @param cr                    Overall information about the client request
     *                              being serviced.
     * @param readOnly              If true, the RatingSection will be read-
     *                              only.  If false, it will be editable.
     */
    public void setReadOnly(ClientRequest cr, boolean readOnly) {
        if (id == null) return;
        cr.evalJavascript("Fiz.ids.@1.readOnly = @2;", id, "" + readOnly);
    }
    
    /**
     * This method is invoked during the final phase of rendering a page; it 
     * generates the necessary HTML and Javascript to make the RatingSection
     * function in the browser.
     * @param cr                    Overall information about the client request
     *                              being serviced.
     */
    @Override
    public void render(ClientRequest cr) {
        Html html = cr.getHtml();
        StringBuilder out = html.getBody();
        
        int numImages = Integer.parseInt(properties.getString("numImages"));
        
        id = properties.containsKey("id") ? 
                properties.getString("id") : cr.uniqueId("rating");

        // Collect the data that will be available for user-specified templates, 
        // including the response to the section's request, if there was one.
        Dataset templateData;
        String requestName = properties.checkString("request");
        if (requestName != null) {
            DataRequest dataRequest = cr.getDataRequest(requestName);
            Dataset response = dataRequest.getResponseOrAbort();
            if (response.containsKey("record")) {
                response = response.getDataset("record");
            }
            templateData = new CompoundDataset(response, cr.getMainDataset());
        } else {
            templateData = cr.getMainDataset();
        }
        
        // Include the necessary authentication tokens and Javascript files.
        cr.setAuthToken();
        html.includeJsFile("static/fiz/RatingSection.js");
        
        // Include the CSS file for RatingSections, assuming the user hasn't
        // opted to do otherwise by specifying their own class.
        if (!properties.containsKey("class")) {
            html.includeCssFile("RatingSection.css");
        }

        // Determine the image family and "unrated" message to use in the 
        // appended HTML.
        String imageFamily = DEFAULT_IMAGE_FAMILY;
        if (properties.containsKey("imageFamily")) {
            imageFamily = properties.getString("imageFamily");
        }
        String unratedMsg = DEFAULT_UNRATED_MESSAGE;
        if (properties.containsKey("unratedMessage")) {
            unratedMsg = Template.expandHtml(properties.getString("unratedMessage"), 
                    templateData);
        }
        
        // Append the HTML for the RatingSection.
        Template.appendHtml(out, "\n<!-- Start RatingSection @1 -->\n" +
                "<div id=\"@1\" class=\"@class?{RatingSection}\" " +
                    "onmousemove=\"@2\" onmouseout=\"@3\" onclick=\"@4\">\n" +
                "  <div id=\"@(1)_offRating\" class=\"offRating\">\n    ", 
                properties, id,
                Template.expandJs("Fiz.ids.@1.mouseMoveHandler(event);", id),
                Template.expandJs("Fiz.ids.@1.mouseOutHandler(event);", id),
                Template.expandJs("Fiz.ids.@1.clickHandler(event);", id));
        for (int i=0; i<numImages; i++) {
            Template.appendHtml(out, 
                    "<img id=\"@(1)_offImg@3\" src=\"@2\" alt=\"Off@3\"></img>",
                    id, StringUtil.addSuffix(imageFamily, OFF_SUFFIX), i + 1);
        }
        Template.appendHtml(out, "\n  </div>\n" +
                "  <div id=\"@(1)_onRating\" class=\"onRating\">\n    ", id);
        for (int i=0; i<numImages; i++) {
            Template.appendHtml(out,  
                    "<img id=\"@(1)_onImg@3\" src=\"@2\" alt=\"On@3\"></img>",
                    id, StringUtil.addSuffix(imageFamily, ON_SUFFIX), i + 1);
        }
        Template.appendHtml(out, "\n  </div>\n" +
                "  <div id=\"@(1)_unratedMsg\" class=\"unratedMsg\">" +
                      "@2?{Not Yet Rated}</div>\n" +
                "</div>\n" + 
                "<!-- End RatingSection @1 -->\n", id, unratedMsg);
        
        // Determine the Ajax URL and initial rating to use in the evaluated 
        // Javascript.
        String ajaxUrl = DEFAULT_AJAX_URL;
        if (properties.containsKey("ajaxUrl")) {
            ajaxUrl = Template.expandUrl(properties.getString("ajaxUrl"), 
                    templateData);
        }
        String initRating = DEFAULT_INIT_RATING;
        if (properties.containsKey("initRating")) {
            initRating = properties.getString("initRating");
        } else if (properties.containsKey("initRatingKey")) {
            initRating = templateData.getString(properties.getString("initRatingKey"));
        }
        
        // Create the initial Javascript to execute when the page is loaded.
        cr.evalJavascript(Template.expandJs("Fiz.ids.@1 = " +
                "new Fiz.RatingSection(\"@1\", @numImages, @granularity?{1}, " +
                "@2, @readOnly?{false}, @autoSizeText?{true}, \"@3\");\n", 
                properties, id, initRating, ajaxUrl));
    }
}

File: web/static/fiz/RatingSection.js
/** RatingSection.js --
 *
 * This file provides Javascript functions needed to implement the
 * RatingSection class.
 *
 * Copyright (c) 2009 Stanford University
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// The following line is used by Fiz to manage Javascript dependencies.
// Fiz:include static/fiz/Fiz.js
// Fiz:include static/fiz/FizCommon.js
// Fiz:include static/fiz/Ajax.js

/*
 * Constructor for the RatingSection.  Takes in a wide variety of information
 * needed by various Javascript functions throughout the existence of of the
 * RatingSection.
 *
 * @param id                The id of the RatingSection {@code <div>}.
 * @param numImages         The number of images displayed in the RatingSection
 *                          (also equal to the maximum possible user rating).
 * @param granularity       The fineness of the rating measurements.  If set to
 *                          zero, then the rating measurement is as fine as the
 *                          pixel grain.
 * @param initRating        The initial rating to display.
 * @param readOnly          If true, then the RatingSection will not respond to
 *                          any triggered events and will therefore be read-
 *                          only.  If false, the RatingSection will update its
 *                          display according to the mouse position and submit
 *                          a rating when the user clicks.
 * @param autoSizeText      If true, then the RatingSection will automatically
 *                          modify the font and positioning of the "unrated"
 *                          message to be centered within the RatingSection.
 *                          If false, the text will be unaltered.
 * @param ajaxUrl           The URL to which an Ajax request should be sent
 *                          upon the user selecting a rating.
 */
Fiz.RatingSection = function(id, numImages, granularity, initRating, readOnly,
        autoSizeText, ajaxUrl) {
    // Save constructor parameters in the object.
    this.numImages = numImages;
    this.granularity = granularity;
    this.readOnly = readOnly;
    this.ajaxUrl = ajaxUrl;
    this.zeroCutoff = 0.25; // The rating will be set to zero if the mouse is
                            // within the left-most 0.25 of the left-most grain.

    // Cache pointers to the key DOM elements.
    this.elemRatingSection = document.getElementById(id);
    this.elemOffRating = document.getElementById(id + "_offRating");
    this.elemOnRating = document.getElementById(id + "_onRating");
    this.elemUnratedMsg = document.getElementById(id + "_unratedMsg");

    // Cache the full width and height of the RatingSection.  In Opera as of
    // 8/2009, only an image's dimensions will always yield the right values.
    // The only exception is if the RatingSection placed in a scrolling element
    // with a width narrower than one image, in which case there (at this time)
    // is no work-around.
    var sampleImg = numImages <= 0 ? {clientWidth: 0, clientHeight: 0} :
        document.getElementById(id + "_offImg1");
    this.fullWidth = Math.max(this.elemOffRating.clientWidth,
        sampleImg.clientWidth * numImages);
    this.fullHeight = Math.max(this.elemOffRating.clientHeight,
        sampleImg.clientHeight);

    // Set the dimensions of the overarching RatingSection {@code <div>}, which
    // will have incorrect dimensions because it isn't absolutely positioned.
    // The unrated message {@code <div>} also won't have the correct dimensions.
    this.elemRatingSection.style.width = this.fullWidth + "px";
    this.elemRatingSection.style.height = this.fullHeight + "px";
    this.elemUnratedMsg.style.width = this.fullWidth + "px";

    // If specified, auto-size the unrated message to be centered within the
    // RatingSection.
    if (autoSizeText) {
        this.elemUnratedMsg.style.fontSize = (this.fullHeight / 2) + "px";
        this.elemUnratedMsg.style.paddingTop = ((this.fullHeight -
            this.elemUnratedMsg.clientHeight) / 2) + "px";
    }

    // Set up the official rating (the most recent submitted or assigned rating)
    // and the shown rating (the rating displayed on the RatingSection).
    this.officialRating = undefined;
    this.shownRating = undefined;

    // Set the initial rating to display.
    this.setOfficialRating(initRating);
}

/*
 * Calculates the rating based on the coordinates of the specified event.
 * Relies heavily on Fiz.getRelativeEventCoords to retrieve the relative
 * x position of the mouse within the RatingSection {@code <div>}.  If a
 * positive non-zero granularity has been specified, the calculated rating is
 * rounded to the nearest granularity (i.e. if granularity is set as 0.5, then
 * the rating will be rounded to the nearest 0.5).  Otherwise, the returned
 * rating will not be rounded.
 *
 * @param event                   The event whose coordinates will determine the
 *                                returned rating.
 * @return                        The calculated rating.
 */
Fiz.RatingSection.prototype.calculateRating = function(event) {
    // Calculate the relative x-position of the event.
    var eventX = Fiz.getRelativeEventCoords(event, this.elemRatingSection).x;
    var rating = -1;

    if (this.granularity > 0) {
        var rawIndex = (eventX * this.numImages) /
            (this.fullWidth * this.granularity);
        // If the mouse is within the left-most zeroCutoff (0.25) of the left-
        // most grain, round the rating to zero.
        var index = (rawIndex <= this.zeroCutoff) ? 0 : Math.ceil(rawIndex);
        rating = Math.min(index * this.granularity, this.numImages);
    } else {
        rating = (eventX * this.numImages) / this.fullWidth;
    }

    return rating;
}

/*
 * Called when the user clicks on the RatingSection.  Sets the currently shown
 * rating as the official rating, and sends that rating via Ajax to the
 * specified URL.
 *
 * @param event                    The triggered click event.
 */
Fiz.RatingSection.prototype.clickHandler = function(event) {
    if (this.readOnly) return;

    this.setOfficialRating(this.shownRating);
    if (this.ajaxUrl != "") {
        new Fiz.Ajax({url: this.ajaxUrl, data:
                {'rating': this.officialRating}});
    }
}

/*
 * Called when the user moves their mouse over the RatingSection.  Calculates
 * the rating based on the coordinates of the specified event, and sets that as
 * the shown rating.
 *
 * @param event                    The triggered mousemove event.
 */
Fiz.RatingSection.prototype.mouseMoveHandler = function(event) {
    if (this.readOnly) return;

    var rating = this.calculateRating(event);
    this.setShownRating(rating);
}

/*
 * Called when the user moves their mouse out of the RatingSection.  Updates
 * the currently shown rating to be the official rating.
 *
 * @param event                    The triggered mouseout or mouseleave event.
 */
Fiz.RatingSection.prototype.mouseOutHandler = function(event) {
    if (this.readOnly) return;

    // If the event was triggered by moving between elements within the
    // RatingSection, do nothing.
    var toElement = event.relatedTarget || event.toElement;
    try {
        if (toElement == this.elemRatingSection ||
            toElement.parentNode == this.elemRatingSection ||
            toElement.parentNode.parentNode == this.elemRatingSection) return;
    } catch (err) {}

    this.setShownRating(this.officialRating);
}

/*
 * Sets the "official" rating of the RatingSection: the official rating
 * represents the most recent rating submitted by the user, or the initial
 * rating if no ratings have been submitted.  If the official and shown ratings
 * are not equal, the shown rating is set to the official rating.  When a user's
 * mouse leaves a RatingSection, the shown rating reverts to the official
 * rating.
 *
 * @param rating                The new official rating.
 */
Fiz.RatingSection.prototype.setOfficialRating = function(rating) {
    this.officialRating = rating;

    if (this.officialRating != this.shownRating) {
        this.setShownRating(this.officialRating);
    }
}

/*
 * Sets the rating shown by the RatingSection; modifies the number or fraction
 * of on/off images to match the specified rating.  If the shown rating is less
 * than zero (implying no rating), then the "unrated" message is displayed.
 *
 * @param rating                The rating to display.
 */
Fiz.RatingSection.prototype.setShownRating = function(rating) {
    this.shownRating = rating;

    if (rating < 0) {
        // If there's no rating, show the unratedMessage and zero images.
        this.elemUnratedMsg.style.display = "block";
        rating = 0;
    } else {
        // If there is a rating, hide the unratedMessage.
        this.elemUnratedMsg.style.display = "none";
    }

    // Show the specified rating by updating the width of elemOnRating.
    var newRatingWidth = rating * this.fullWidth / this.numImages;
    this.elemOnRating.style.width = newRatingWidth + "px";
}
